aten/src/ATen/SparseTensorImpl.h

```diff
@@ -1,10 +1,83 @@
 #pragma once
 
+#include "ATen/Tensor.h"
 #include "ATen/TensorImpl.h"
+#include "ATen/Error.h"
 
 namespace at {
 struct SparseTensorImpl : public TensorImpl {
-  explicit SparseTensorImpl(Type * type)
-  : TensorImpl(type) {}
+  // Stored in COO format, indices + values
+
+  std::vector<int64_t> size_;
+  // INVARIANT: indices_.size(1) >= nnz_ && values_.size(0) >= nnz_
+  int64_t nnz_ = 0;
+  int64_t dimI_ = 0; // number of sparse dimensions
+  int64_t dimV_ = 0; // number of dense dimensions
+
+  // 2-D tensor of nDim x nnz of indices. May have nnz dim bigger than nnz
+  // as buffer, so we keep track of both
+  Tensor indices_; // always a LongTensor
+  Tensor values_;
+
+  // A sparse tensor is 'coalesced' if every index occurs at most once in
+  // the indices tensor, and the indices are in sorted order.
+  // Most math operations can only be performed on ordered sparse tensors
+  bool coalesced_ = false;
+
+public:
+  // Public for now...
+  explicit SparseTensorImpl(Type * type);
+
+  int64_t nnz() const { return nnz_; }
+  int64_t dimI() const { return dimI_; }
+  int64_t dimV() const { return dimV_; }
+  bool coalesced() const { return coalesced_; }
+  Tensor indices() const { return indices_; }
+  Tensor values() const { return values_; }
+
+  const char * toString() const override;
+  IntList sizes() const override;
+  IntList strides() const override;
+  int64_t dim() const override;
+  Scalar localScalar() override;
+  void * unsafeGetTH(bool retain) override;
+  std::unique_ptr<Storage> storage() override;
+
+  // Some ops do some manual size fiddling.
+  // TODO: Figure out a more safe way to provide this functionality
+  std::vector<int64_t>& _sizes_mut() { return size_; }
+
+  // WARNING: This function does NOT preserve invariants of dimI/dimV with
+  // respect to indices and values
+  void raw_resize_(int64_t dimI, int64_t dimV, ArrayRef<int64_t> size) {
+    // UGHHHHH.  Legacy special case
```

Comment:
you could future proof this code with USE_TH_ZERO_SIZE_DIM as in https://github.com/pytorch/pytorch/pull/8468 if you want.

Reply:
sure, if you get the tests to pass before I merge this XD